build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -229,11 +251,43 @@ service ActionCache {
 // `INVALID_ARGUMENT` error will be returned. In either case, the client should
 // not attempt to retry the upload.
 //
-// For downloading blobs, the client must use the
-// [Read method][google.bytestream.ByteStream.Read] of the ByteStream API, with
-// a `resource_name` of `"{instance_name}/blobs/{hash}/{size}"`, where
-// `instance_name` is the instance name (see above), and `hash` and `size` are
-// the [Digest][build.bazel.remote.execution.v2.Digest] of the blob.
+// Small downloads can be grouped and requested in a batch via
+// [BatchReadBlobs][build.bazel.remote.execution.v2.ContentAddressableStorage.BatchReadBlobs].
+//
+// For large downloads, the client must use the
+// [Read method][google.bytestream.ByteStream.Read] of the ByteStream API.
+//
+// For uncompressed data, The `ReadRequest.resource_name` is of the following form:
+// `{instance_name}/blobs/{hash}/{size}`
+// Where `instance_name`, `hash` and `size` are defined as for uploads.
+//
+// Data can alternatively be downloaded in compressed form, with the following
+// `ReadRequest.resource_name` form:
+// `{instance_name}/compressed-blobs/{compressor}/{uncompressed_hash}/{uncompressed_size}`
+//
+// Where:
+// * `instance_name` and `compressor` are defined as for uploads.
+// * `uncompressed_hash` and `uncompressed_size` refer to the
+//   [Digest][build.bazel.remote.execution.v2.Digest] of the data being
+//   downloaded, once uncompressed. Clients MUST verify that these match
```

Comment:
I can understand making verification a requirement for the server, because clearly specifying the error behavior is necessary. Is it really an API requirement that clients verify the data? What does it defend against?

Reply:
I think I added this in response to your earlier issue: https://github.com/bazelbuild/remote-apis/pull/168/commits/d342599f7e7713f031ee90f900ea7bc6d2994937#r483648050 "The same rules about verifying data and retrying apply here, too". In commit ed21a18ba9ff0acfc4c6cac34d4c8df591fa4297 this text was changed from "Clients SHOULD" to "Clients MUST". Perhaps I misunderstood your earlier comment?

Verifying these hashes protects against data corruption (on disk, or in transit, or perhaps in compression).

It is undesirable for servers to perform this check, because that would require that they decompress + hash these compressed blobs- whereas one of the goals of this feature is to avoid this work and stream compressed blobs directly.

Clients on the other hand need to decompress this data anyway, and if servers aren't verifying then I think the clients should be. I'm not sure if they "must", though.